Cram embed ref cons

jkbonfield opened the pull request:
A better implementation of embed_ref=2 now I've realised I was mistaken in the main code structure.  It's much easier than I thought to do a proper consensus calculation,  so I now use this.  This means:

- With MD:Z tags, the embed_ref=2 (implicit) is the same as embed_ref=1 (explicit via external .fa file).
- Without MD:Z tags, embed_ref=2 is now usually slightly smaller than an explicit embed_ref=1 as the consensus is always at least as accurate as the official reference.  The old code was sometimes a bit suspect in the generated reference.

This also addresses some of the issues raised in https://github.com/samtools/htslib/issues/1445, for better or worse.  My decisions to the questions raised there were:

1.  "If we use the embed_ref option, should it automatically use level 2 if no external reference is found?"
Yes.  An unadorned  "embed_ref" is still level 1, but when a reference cannot be found it switches to level 2.  Note it does this on the file descriptor so it applies to all containers.

2. "If an external reference is not found during encode and no embed_ref option has been set, should we automatically switch to embed_ref=2 mode? "
Yes, for both when MD:Z are known and when not, as it's not easy to know whether they are present, and there are middle-ground cases such as partially present, or present but incorrect so it's not good to try and separate logic by content.  A warning message is emitted stating that the file will likely be larger than when using an external reference.

3. "Should we be providing some hint mechanism in the SAM header, eg by adding @CO htslib_cram_options:embed_ref and similar."
Currently no.  With the notion of simply auto-generating an embedded reference when one doesn't exist, this is less required anyway.

Note this isn't a complete panacea.  Unsorted data cannot use embedded references, and neither can multi_seq_per_slice mode (eg when encoding a highly fragmented assembly).

Thoughts?

jkbonfield commented:
Bah I found another corner case.

The SAM header must have M5 tags when outputting in CRAM, unless an explicit embed_ref=2 or no_ref is used.  An example is the classic NA12878_S1.bam from Illumina.  This has no meta-data whatsoever; just SQ SN and LN fields leaving the user totally guessing if it's GRCh37, GRCh38, or IIRC it's actually NCBI's 37 with a subtly different MT genome (sigh).

I'm not sure how best to handle that.  It's just another hard failure right now, but with an explicit embed_ref=2 it works fine and there are MD:Z tags present so it'll even be efficient for retaining those.

daviesrob commented on cram/cram_encode.c:
```diff
@@ -49,6 +49,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include "../sam_internal.h" // for nibble2base
 #include "../htslib/hts.h"
 #include "../htslib/hts_endian.h"
+#include "../textutils_internal.h"
```
This needs to be updated in the Makefile.

jkbonfield commented on cram/cram_encode.c:
```diff
@@ -1401,6 +1402,286 @@ static int add_read_names(cram_fd *fd, cram_container *c, cram_slice *s,
+// Returns the next cigar op code: one of the BAM_C* codes,
+// or -1 if no more are present.
+static inline
+int next_cigar_op(uint32_t *cigar, int *ncigar, int *skip, int *spos,
```
Regarding "skip", it could be moved there, but the `next_cigar_op` is a general purpose function (albeit static), with skip being a parameterisation of it.  I could move it there, but then it'd need renaming to something less general.  I doubt it matters much so inclined to leave as-is for now.

jkbonfield commented on cram/cram_encode.c:
```diff
@@ -1401,6 +1402,286 @@ static int add_read_names(cram_fd *fd, cram_container *c, cram_slice *s,
+    // Otherwise we just use SEQ+CIGAR and build a consensus which we later
+    // turn into a fake reference
+    uint32_t *cigar = bam_get_cigar(b);
+    int ncigar = b->core.n_cigar;
+    int i, j;
```
I see there is an `ncigar` field in the internal cram record struct too, which is signed.  Sloppy! (Sorry)

I'll look into these as a separate issue though.

jkbonfield commented on cram/cram_encode.c:
```diff
@@ -1442,8 +1723,20 @@ int cram_encode_container(cram_fd *fd, cram_container *c) {
+                if (fd->embed_ref <= 0)
+                    hts_log_warning("NOTE: the CRAM file will be bigger than"
+                                    " using an external reference");
+                pthread_mutex_lock(&fd->ref_lock);
```
Good idea for putting into a local variable, as is already done just about with `nref`.  Regarding elsewhere (such as `process_read`) I did think about locks but the various automated tools seemed to think it was impossible to trigger data races there.

I'll explore again with more tools.  It's possible it can still happen if we fail to find a reference part way through a file instead of only at the first one (which essentially means it hasn't yet dispatched a single thread job, hence impossible to trigger).  So I'll stress test more avenues.

jkbonfield commented:
I think this should be working now.  Thanks for the review.  It can all be squashed together when you're OK with it.

jkbonfield commented:
Added a commit to move `fd->embed_ref` to local variables / arguments, so one container changing the global default doesn't affect the running behaviour of code in another thread.

Also rebased so the tests should now pass.

JosephLalli commented:
@jkbonfield 
Question: Is there potential for an option to embed short references, or to specify more than one reference file? I am working with a dataset of reads aligned to GRCh38 + several thousand 100-100,000 bp-long contigs of sample-specific insertions. An ideal situation would be a CRAM file that refers to a common external reference genome, but can also use a separate sample-specific reference for small contigs (either via a sample-specific fasta or with embedded small contigs).

If this is already possible, then I apologize! I've looked around and can't find anything.
